homeassistant/components/alexa/smart_home.py

```diff
@@ -927,12 +966,55 @@ def async_setup(hass, config):
     Even if that's disabled, the functionality in this module may still be used
     by the cloud component which will call async_handle_message directly.
     """
+
     smart_home_config = Config(
+        client_id=config[CONF_CLIENT_ID],
+        client_secret=config[CONF_CLIENT_SECRET],
         should_expose=config[CONF_FILTER],
         entity_config=config.get(CONF_ENTITY_CONFIG),
     )
     hass.http.register_view(SmartHomeView(smart_home_config))
 
+    global AUTH
+    AUTH = Auth(hass, smart_home_config.client_id,
+                smart_home_config.client_secret)
+
+    hass.loop.create_task(async_enable_proactive_mode(hass, smart_home_config))
+
+
+async def async_enable_proactive_mode(hass, smart_home_config):
+    if not await AUTH.async_is_auth_done():
+        # not ready yet
+        return
+
+    @ha.callback
+    def async_entity_state_listener(changed_entity, old_state, new_state):
+        if not smart_home_config.should_expose(changed_entity):
+            _LOGGER.debug("Not exposing %s because filtered by config",
+                          changed_entity)
+            return
+
+        entity = hass.states.get(changed_entity)
+        if not entity:
+            return
+
+        if entity.domain not in ENTITY_ADAPTERS:
+            return
+
+        alexa_changed_entity = \
+            ENTITY_ADAPTERS[entity.domain](hass, smart_home_config, entity)
+
+        for interface in alexa_changed_entity.interfaces():
+            if interface.properties_proactively_reported():
+                # TODO: we should check if this entity was already discovered
+                # by Alexa (store them in the prefs?) and, if not, send a
+                # proactive discovery message instead.
+                hass.async_add_job(async_send_changereport_message, hass,
```

Comment:
I am confused, we're not passing in the interface, just the entity. So for each interface that is reported, we report on the entity multiple times?

We should also not use `async_add_job` but, if we really need multiple tasks, create them but `await asyncio.wait(…)` them, so that stacktraces can lead to the right piece in the code.

Reply:
There is a `return` on the next line so that it will only report once for each entity.

Reply:
In that case we should just make this whole method async and await the changereport sending method instead of adding it as a job.

Reply:
If I remember correctly, that was my first approach, but it wouldn't work since the `async_entity_state_listener` has to be a callback and could not be async. I've done that a while ago, so I may be wrong.

Reply:
`async_track_state_change` can handle async, callback and normal methods.

Reply:
Not sure why this is resolved. We should `await async_send_changereport_message(…)` here and mark `async_entity_state_listener` as async.